util/rate_limiter.cc

```diff
@@ -324,6 +320,301 @@ Status GenericRateLimiter::Tune() {
+  for (int q = 0; q < 2; ++q) {
+    auto use_pri = (use_low_pri_first == q) ? Env::IO_LOW : Env::IO_HIGH;
+    auto* queue = &queue_[use_pri];
+    while (!queue->empty()) {
+      auto* next_req = queue->front();
+      if (available_bytes_ < next_req->request_bytes) {
+        // avoid starvation
+        next_req->request_bytes -= available_bytes_;
+        available_bytes_ = 0;
+        break;
+      }
+      available_bytes_ -= next_req->request_bytes;
+      next_req->request_bytes = 0;
+      total_bytes_through_[use_pri] += next_req->bytes;
+      duration_bytes_through_ += next_req->bytes;
+      queue->pop_front();
+
+      next_req->granted = true;
+      if (next_req != leader_) {
+        // Quota granted, signal the thread
+        next_req->cv.Signal();
+      }
+    }
+  }
+}
+
+int64_t GenericRateLimiterV2::CalculateRefillBytesPerPeriod(
+    int64_t rate_bytes_per_sec) {
+  if (port::kMaxInt64 / rate_bytes_per_sec < refill_period_us_) {
+    // Avoid unexpected result in the overflow case. The result now is still
+    // inaccurate but is a number that is large enough.
+    return port::kMaxInt64 / 1000000;
+  } else {
+    return std::max(kMinRefillBytesPerPeriod,
+                    rate_bytes_per_sec * refill_period_us_ / 1000000);
+  }
+}
+
+Status GenericRateLimiterV2::Tune() {
+  const int kAllowedRangeFactor = 20;
+  // high-priority bytes are padded to 1MB
+  const int64_t kHighBytesLower = 1024 * 1024;
+  // lower bound for write amplification estimation
+  const int kRatioLower = 15;
+  // since compaction cannot fully utilize the IO quota we gave, use ratio
+  // slightly bigger than estimation to adjust target limit.
+  const int kRatioPadding = 4;
+
+  std::chrono::microseconds prev_tuned_time = tuned_time_;
+  tuned_time_ = std::chrono::microseconds(NowMicrosMonotonic(env_));
+  auto duration = tuned_time_ - prev_tuned_time;
+  auto duration_ms =
+      std::chrono::duration_cast<std::chrono::milliseconds>(duration).count();
+
+  int64_t prev_bytes_per_sec = GetBytesPerSecond();
+
+  bytes_sampler_.AddSample(duration_bytes_through_ * 1000 / duration_ms);
+  highpri_bytes_sampler_.AddSample(duration_highpri_bytes_through_ * 1000 /
+                                   duration_ms);
+  duration_bytes_through_ = 0;
+  duration_highpri_bytes_through_ = 0;
+  // in case there are compaction burst even when online writes are stable
+  auto util = bytes_sampler_.GetRecentValue() * 100 / prev_bytes_per_sec;
+  if (util >= 98) {
+    ratio_delta_ += 1;
+  } else if (util < 90 && ratio_delta_ > 0) {
+    ratio_delta_ -= 1;
+  }
+  int32_t ratio = std::max(
+      kRatioLower,
+      static_cast<int32_t>(
+          bytes_sampler_.GetFullValue() * 10 /
+          std::max(highpri_bytes_sampler_.GetFullValue(), kHighBytesLower)));
+  int64_t new_bytes_per_sec =
+      (ratio + kRatioPadding + ratio_delta_) *
+      std::max(highpri_bytes_sampler_.GetRecentValue(), kHighBytesLower) / 10;
```

Comment:
write burst may cause higher bytes_per_sec. but as I understand it'd better slower the compaction at that time.

Reply:
There are two cases worth considering. First is that you wish to have steady IO even then flush flow is jittering. I give you smooth window (`RecentValue`) to shield against jitter within 10 second.
Second is when physical resources are drained, you would like to assign more IO quota to flush since it's of higher priority. But I think in the long run it's unwise to do so because sooner or later where will be write stalls. Instead the limiter grants more quotas to compaction and jamming the IO traffic, flush and compaction are both slowed down gracefully.

Reply:
I think as long as `max_bytes_per_sec` is not reached, we can afford letting compaction do more work to avoid write stalls. In another word, the goal of the auto-tune algorithm (both the new one and the old one) is smoothing compaction flow, not protecting frontend write performance before `max_bytes_per_sec` is hit.